stan/math/rev/core/vari.hpp

```diff
@@ -27,21 +64,33 @@ class var;
  * classes will store operand variables and propagate derivative
  * information via an implementation of chain().
  */
-class vari {
+template <typename T, typename = void>
+class vari_value : public vari_base {};
+
+template <typename T>
+class vari_value<T, std::enable_if_t<std::is_arithmetic<T>::value>>
+    : public vari_base {
  private:
-  friend class var;
+  template <typename>
+  friend class var_type;
+  template <typename Val>
+  using floating_point_promoter
+      = std::conditional_t<std::is_integral<std::decay_t<Val>>::value, double,
+                           std::decay_t<Val>>;
 
  public:
+  using Scalar = floating_point_promoter<T>;
+  using value_type = Scalar;
   /**
    * The value of this variable.
    */
-  const double val_;
+  const Scalar val_;
```

Comment:
For this to not cause memory leaks, `Scalar` needs to be something that does not require its destructor to be call. Somehow that needs documated.

In this theme, should we try to make this templated thing as general as possible? Or should we just implement: `vari_matrix`, `vari_vector`, `vari_row_vector` and use those everywhere? (they could certainly depend on something templated)

Reply:
So this specialization is just for arithmetic types

```cpp
class vari_value<T, std::enable_if_t<std::is_arithmetic<T>::value>>
    : public vari_base {
```

It should do the same thing as the previous version. Will having two arithmetic types here leak memory?

Reply:
The actual Eigen implementation is here. Do you want me to open up a PR on that?

https://github.com/stan-dev/math/compare/feature/eigen-vari#diff-130c5a75cc427d7d41715e9fca8281f4R176

Reply:
> std::enable_if_t<std::is_arithmetic<T>::value>

Oooh, got it.

> The actual Eigen implementation is here.

Oooh, I see.

> Do you want me to open up a PR on that?

Yeah, this makes me want to do it all together. I don't think I'm gonna be able to keep all this stuff separate in my head and I'm just gonna keep asking questions that have been answered elsewhere.

Reply:
Word! See below

https://github.com/stan-dev/math/pull/1877